enhancements/machine-api/spot-instances.md

```diff
@@ -0,0 +1,359 @@
+A DaemonSet should be deployed on the cluster which runs Pods on nodes labelled `machine.openshift.io/terminate-on-preemption: true`.
+Users of Spot instances could optionally add this label to their Machine’s spec to opt in to this feature.
+
+The DaemonSet itself would poll the EC2 metadata service for a termination notice.
+As soon as it receives an `OK` response (a non-terminated instance returns a `404` response), it would delete the Machine object associated with the node that it is running on.
+This deletion will trigger the Machine controller to start draining the node, ready for its termination.
+
+##### GCP
+
+###### Launching instances
+
+To launch an instance as Preemptible on GCP, the `Preemptible` field must be set:
+
+```
+&compute.Instance{
+  ...
+  Scheduling: &comput.Scheduling{
+    ...
+    Preemptible: true,
+  },
+}
+```
+
+Therefore, to make the choice up to the user, this field should be added to the `GCPMachineProviderSpec`:
+
+```
+type GCPMachineProviderSpec struct {
+  ...
+  Preemptible bool `json:”preemptible”`
+}
+```
+
+##### Azure
+
+###### Launching Instances
+
+To launch a VM as a Spot VM on Azure, the following 3 options need to be set within the [VirtualMachineProperties](https://github.com/Azure/azure-sdk-for-go/blob/8d7ac6eb6a149f992df6f0392eebf48544e2564a/services/compute/mgmt/2019-07-01/compute/models.go#L10274-L10309)
+when the instance is created:
+
+- Priority: This must be set to `Spot` to request a Spot VM
+
+- Eviction Policy: This has two options, `Deallocate` or `Delete`. Only `Deallocate` is valid when using Spot VMs and as such, this must be set to `Deallocate`.
+
+- BillingProfile (default: -1) : This is a struct containing a single field, `MaxPrice`.
+  This is a float representation of the maximum price the user wishes to pay for their VM.
+  This defaults to -1 which makes the maximum price the On-Demand price for the instance type.
+  This also means the instance will never be evicted for price reasons as Azure caps Spot Market prices at the On-Demand price.
+
+The only option that a user needs to interact with is the `MaxPrice` field within the `BillingProfile`, other fields only have 1 valid choice and as such can be inferred.
+Similar to AWS, we can make an optional struct for SpotVMOptions, which, if present, implies the priority is `Spot`.
+
+```
+type SpotVMOptions struct {
+  MaxPrice *float64 `json:”maxPrice,omitempty”`
+}
+
+type AzureMachineProviderSpec struct {
+  ...
+
+  SpotVMOptions *SpotVMOptions `json:”spotVMOptions,omitempty”`
+}
+```
+
+###### Deallocation
+Since Spot VMs are not deleted when they are preempted and instead are deallocated,
+we will need to monitor for deallocated instances and manually cleanup the leftover resources (VM, disks, networking).
+A MachineHealthCheck on Spot VMs should be capable of noticing that the machine has stopped and initiating a delete on the VM by the Machine actuator.
+
+### Risks and Mitigations
+
```

Comment:
We need to add some best-practices for the autoscaler.  Things might go a little bit crazy if the cluster autoscaler is turned on and is not spot-aware.